develop/hashing/main.cpp

```diff
@@ -219,6 +220,165 @@ struct quadratic_table {
+        int value;
+        bool valid = false;
+    };
+
+    bucket_cuckoo_table()
+        : cells{new cell[(M)]{}},
+        as{new int[(d)]{}},
+        bs{new int[(d)]{}}
+        {
+            rerollHashFunctions();
+        }
+
+    int randomOddInt(){
+        int r = chooseHash(prng);
+        return r&1?r:r+1;
+    }
+    void rerollHashFunctions() {
+        for (int i = 0; i < d; ++i) {
+            as[i] = randomOddInt();
+            bs[i] = randomOddInt();
+        }
+    }
+
+    int hash_to_bucket_index(int h) {
+        return (h & ((M >> log2B) - 1)) * B;
+    }
+
+    //d = hash functions
+    //B = buckets => linear probe through buckets
+    //hash table consists of m/B buckets - each containing B cells => linear probe B times per Bucket
+    //Cuckoo hash functions h1, ..., hd select d bucket indices.
+
+    bool rehash(int currentKey, int currentValue){
+        rerollHashFunctions();
+        cell *tmp = new cell[M];
+        std::copy(cells, cells + M, tmp);
+        delete[] cells;
+        cells = nullptr;
+        cells = new cell[M];
+        //reinsert all other key-value-pairs
+        bool succeeded;
+        for(int i = 0; i<M; ++i){
+            auto &c = cells[i];
+            if(c.valid){
+                succeeded = putHelper(c.key, c.value, 0, false);
+                if(!succeeded){
+                    delete[] cells;
+                    cells = nullptr;
+                    cells = new cell[M];
+                    std::copy(tmp, tmp + M, cells);
+                    delete[] tmp;
+                    tmp = nullptr;
+                    return true;
+                }
+            } else {
+                //nothing to insert
+            }
+        }
+        //reinsert current key-value-pair
+        if(succeeded) {
+            if(!putHelper(currentKey, currentValue, 0, false)){
+                delete[] cells;
+                cells = nullptr;
+                cells = new cell[M];
+                std::copy(tmp, tmp + M, cells);
+                delete[] tmp;
+                tmp = nullptr;
+                return true;
+            };
+        }
+
+        return false;
+    }
+
+    bool putHelper(int k, int v, int chain = 0, bool hashAgain = true) {
+        if(chain > max_eviction_length) {
+            bool rehashSuccess = true;
+            while(hashAgain && rehashSuccess) {
+                rehashSuccess = rehash(k, v);
+            }
+            if(!rehashSuccess){
+                return true;
+            }
+            return false;
+        }
+        for(int i = 0; i < d; i++){
+            auto idx = hash_to_bucket_index((as[i] * (unsigned)k + bs[i]) >> (32 - 26));
+
+            for(int i2 = 0; i2 < B; i2++){
+                auto &c = cells[idx+i2];
+                if(!c.valid) {
+                    c.key = k;
+                    c.value = v;
+                    c.valid = true;
+                    return true;
+                }
+
+                if(c.key == k) {
```

Comment:
this is wrong behaviour. I will first have to check **ALL** possible cells for occurrence. And if the key does not occur then I am allowed to insert into an empty cell (if there is one).